node/cn/api_sc_backend_test.go

```diff
@@ -0,0 +1,402 @@
+	api.SetHead(headNumber)
+}
+
+func TestServiceChainAPIBackend_GetBlock(t *testing.T) {
+	{
+		api, mockCtrl, mockBlockChain, _ := prepareServiceChainAPIBackendTest(t)
+		mockBlockChain.EXPECT().GetBlockByHash(hashes[0]).Return(nil).Times(1)
+
+		block, err := api.GetBlock(context.Background(), hashes[0])
+		assert.Nil(t, block)
+		assert.Error(t, err)
+		mockCtrl.Finish()
+	}
+	{
+		api, mockCtrl, mockBlockChain, _ := prepareServiceChainAPIBackendTest(t)
+		mockBlockChain.EXPECT().GetBlockByHash(hashes[0]).Return(blocks[0]).Times(1)
+
+		block, err := api.GetBlock(context.Background(), hashes[0])
+		assert.Equal(t, blocks[0], block)
+		assert.NoError(t, err)
+		mockCtrl.Finish()
+	}
+}
+
+func TestServiceChainAPIBackend_GetTxAndLookupInfo(t *testing.T) {
+	api, mockCtrl, mockBlockChain, _ := prepareServiceChainAPIBackendTest(t)
+	defer mockCtrl.Finish()
+
+	expectedTx := tx1
+	expectedHash := hashes[0]
+	expectedBlockNum := uint64(123)
+	expectedIndex := uint64(321)
+
+	mockBlockChain.EXPECT().GetTxAndLookupInfo(hashes[0]).Return(tx1, hashes[0], uint64(123), uint64(321)).Times(1)
+
+	tx, hash, blockNum, index := api.GetTxAndLookupInfo(hashes[0])
+
+	assert.Equal(t, expectedTx, tx)
+	assert.Equal(t, expectedHash, hash)
+	assert.Equal(t, expectedBlockNum, blockNum)
+	assert.Equal(t, expectedIndex, index)
+}
+
+func TestServiceChainAPIBackend_GetBlockReceipts(t *testing.T) {
+	api, mockCtrl, mockBlockChain, _ := prepareServiceChainAPIBackendTest(t)
+	defer mockCtrl.Finish()
+
+	rct1 := newReceipt(111111)
+	rct2 := newReceipt(222222)
+	rct3 := newReceipt(333333)
+
+	receipts := types.Receipts{rct1, rct2, rct3}
+
+	mockBlockChain.EXPECT().GetReceiptsByBlockHash(hashes[0]).Return(receipts).Times(1)
+	returnedReceipts := api.GetBlockReceipts(context.Background(), hashes[0])
+
+	assert.Equal(t, receipts, returnedReceipts)
+}
+
+func TestServiceChainAPIBackend_GetLogs(t *testing.T) {
+	api, mockCtrl, mockBlockChain, _ := prepareServiceChainAPIBackendTest(t)
+	defer mockCtrl.Finish()
+
+	expectedLogs := [][]*types.Log{{{BlockNumber: 123}, {BlockNumber: 321}}}
+	mockBlockChain.EXPECT().GetLogsByHash(hashes[0]).Return(expectedLogs).Times(1)
+
+	returnedLogs, err := api.GetLogs(context.Background(), hashes[0])
+	assert.Equal(t, expectedLogs, returnedLogs)
+	assert.NoError(t, err)
+}
+
+func TestServiceChainAPIBackend_GetTd(t *testing.T) {
+	api, mockCtrl, mockBlockChain, _ := prepareServiceChainAPIBackendTest(t)
+	defer mockCtrl.Finish()
+
+	expectedTD := big.NewInt(12345)
+	mockBlockChain.EXPECT().GetTdByHash(hashes[0]).Return(expectedTD).Times(1)
+	assert.Equal(t, expectedTD, api.GetTd(hashes[0]))
+}
+
+func TestServiceChainAPIBackend_SubscribeRemovedLogsEvent(t *testing.T) {
+	api, mockCtrl, mockBlockChain, _ := prepareServiceChainAPIBackendTest(t)
+	defer mockCtrl.Finish()
+
+	ch := make(chan<- blockchain.RemovedLogsEvent, 10)
+
+	mockBlockChain.EXPECT().SubscribeRemovedLogsEvent(ch).Times(1)
+	api.SubscribeRemovedLogsEvent(ch)
```

Comment:
Is the purpose of this test to verify that no panic or hang occurs?

Reply:
It seems that I missed checking the return variable of functions. Thanks! I will update the tests.

Reply:
@jeonghonah After take a deeper look into the test, the purpose of these tests are to check if only the desired calls to the mockBlockChain happens or not. So it looks fine as it is.